app/controllers/users/registrations_controller.rb

```diff
@@ -1,10 +1,41 @@
-# frozen_string_literal: true
-
 class Users::RegistrationsController < Devise::RegistrationsController
+  alias :rebuild_devise_user :build_resource
```

Comment:
I would personally have moved this to the model, as it is a `User` validation, and used the Rails `validate` method.

You can either use a method to validate: https://guides.rubyonrails.org/active_record_validations.html#custom-methods or a validator class: https://guides.rubyonrails.org/active_record_validations.html#custom-validators

Then we shouldn't need to override this controller at all, and the validator will add the errors to the unsaved `User` object and set the error in the form.

Reply:
I thought about this, but I felt like tying us to a ActiveModel-specific way of doing it might be more pain as this whitelist becomes more complicated. 

Simple validations - making sure something isn't blank - seem fine to sit there, it wouldn't be surprising to the next reader to find them there, and . This one seems more like a _business rule_ than a _validation_ to me, and I like pulling it up out of the model to make it more obvious because of that. I envisioned it being a `use case` in the near-future.

I'm still going back and forth on "if we're using Rails, may as well go full Rails" and "must keep as little in the framework as possible". Be interested to hear what rest of team thinks.

Reply:
Hmm, I still think it belongs on the User model. To me its not a business case its Model validation: "a user cannot be created if it doesn't have a suitable email address". The controller is not a great place for this logic if we are sticking to MVP (regardless of the methods/techniques rails offers).

If this does get complicated, you can still use a `UseCase` for sure, but it should still be called from the Model, eg: `validate :confirm_email_user_case`, where the `def confirm_email_user_case` would call the `UseCase` and return `true` or `false`. But importantly, it belongs on the user because it determines if the user can be created or not.

Reply:
Talked about this in person - validator sounds good!